File: LeetCode/BinaryTreeDFS/JavaScript/sameTree.js
// Given the roots of two binary trees p and q, write a function to check if they are the same or not.

// Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

// Example 1:

// Input: p = [1,2,3], q = [1,2,3]
// Output: true
// Example 2:

// Input: p = [1,2], q = [1,null,2]
// Output: false
// Example 3:

// Input: p = [1,2,1], q = [1,1,2]
// Output: false

// Constraints:

// The number of nodes in both trees is in the range [0, 100].
// -104 <= Node.val <= 104

// TC: O(max(p,q)) We process the left and right children of each nodes starting from the root. We have 2 roots.
// SC: O(max number of nodes on a level of either p or q). We add each children node to a stack for processing as a pair.

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function (p, q) {
  // Using Iterative DFS approach
  // Since JS do not have an XOR conditional operator, we have to create our own
  // XOR Truth table
  //      0  1
  //     ------
  //  0 | 0  1
  //  1 | 1  0
  //
  const xor = (a, b) => Boolean(!a ^ !b);

  // Let's handle the edge case, return false if only either p or q is null
  if (xor(p, q)) {
    return false;
  }

  // If both roots are null, then they're the same
  if (!p && !q) {
    return true;
  }

  // We can use BFS or DFS and visit each nodes starting at the two roots.
  // Using Iterative DFS
  let stack = []; // we can push both roots on the stack as a pair

  stack.push([p, q]); // [remember that the order of the pair is p then q]

  while (stack.length) {
    // Pop left from the stack
    let [nodeP, nodeQ] = stack.pop();

    // At this point, we're guaranteed that both nodes are non null, so, we check if the values are equal
    if (nodeP.val !== nodeQ.val) {
      return false;
    }

    // Check if the children nodes are present. We return false if either the p or q child is null
    if (xor(nodeP.left, nodeQ.left)) {
      return false;
    }
    if (xor(nodeP.right, nodeQ.right)) {
      return false;
    }

    // At this point, the children nodes are both non-null or null, we add it to the stack if they're non-null
    if (nodeP.left && nodeQ.left) stack.push([nodeP.left, nodeQ.left]);
    if (nodeP.right && nodeQ.right) stack.push([nodeP.right, nodeQ.right]);
  }

  return true;
};

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function (p, q) {
  // Using Recursive DFS approach

  // Base cases
  if (!p && !q) {
    return true;
  }

  // We return false if one of the node is null or if the node values are not equal
  if (!p || !q || p.val !== q.val) {
    return false;
  }

  // At each recursion, we check if both trees are equal, so we compare the right from the left and both should be true
  return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
};

File: LeetCode/BinaryTreeDFS/python/sameTree.py
# Definition for a binary tree node.
from collections import deque
from typing import Dict, List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # Using recursive DFS
        # base case
        if p is None and q is None:
            return True
        
        # check node if equal
        if p is None or q is None or p.val != q.val:
            return False

        # At each recursion, we check if both trees are equal, so we compare the right
        # from the left and both should be true
        return (self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right))

    @staticmethod
    def testSolution(input: Dict[str, List[str | int]]) -> None:
        # Create the binary tree using BFS
        pBinTree = Solution.BFSCreateBinaryTree(input["p"])
        qBinTree = Solution.BFSCreateBinaryTree(input["q"])

        # Print the binary trees
        print("Input: p: " + Solution.BFSBinaryTreeToStr(pBinTree))
        print("Input: q: " + Solution.BFSBinaryTreeToStr(qBinTree))
        print("Expected: " + str(input["expected"]))

        # Call the function to be tested
        res = Solution().isSameTree(pBinTree, qBinTree)

        # Print the result
        print("Result: " + str(res))

        # Validate the result
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)

    @staticmethod
    def BFSCreateBinaryTree(listVal: List[str | int]) -> TreeNode:
        arLen = len(listVal)

        if arLen < 1:
            return None
        
        root =  TreeNode(listVal[0], None, None)
        i = 1

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])

        while len(queue) > 0 and i < arLen:
            curNode = queue.popleft()

            if curNode is not None:
                # left child
                if i < arLen and listVal[i] is not None:
                    curNode.left = TreeNode(listVal[i], None, None)
                    queue.append(curNode.left)
                i += 1

                # right child
                if i < arLen and listVal[i] is not None:
                    curNode.right = TreeNode(listVal[i], None, None)
                    queue.append(curNode.right)
                i += 1

        return root
    
    @staticmethod
    def BFSBinaryTreeToStr(root: Optional[TreeNode]) -> str:
        if root is None:
            return "[ ]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            tmp = str(curNode.val) if curNode is not None else "null"

            sb.append(tmp)
 
            if curNode is not None:
                queue.append(curNode.left)
                queue.append(curNode.right)

        new_len = 0

        # Check starting from end of sb_array where the value is not "null"
        for i in range(len(sb) - 1, -1, -1):
            if sb[i] == "null":
                continue
            else:
                new_len = i
                break

        # Remove those extra "null"
        if len(sb) > 0:
            sb = sb[0 : new_len + 1]

        return "[ " + ", ".join(sb) + " ]"

if __name__ == "__main__":
    input = {"p": [1,2,3], "q": [1,2,3], "expected": True}
    Solution.testSolution(input)

    input = {"p": [1,2], "q": [1,"null",2], "expected": False}
    Solution.testSolution(input)

    input = {"p": [1,2,1], "q": [1,1,2], "expected": False}
    Solution.testSolution(input)